### PR TITLE
Port VV stuff from TG

### PR DESCRIPTION
**Pull request by Xander3359**
## About The Pull Request
Ports VV stuff from TG
image

## Why It's Good For The Game
Better admin 👍 
Better code 👍 
Better life 👍 
## Changelog
:cl:
code: Port VV Stuff from TG
admin: osay
admin: VV Panel should be fully functional with new options in the dropdown menu
admin: Added tagging/marking datums from TG
/:cl:

### CODE REVIEW

File: code/modules/admin/view_variables/topic.dm [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	if( (usr.client != src) || !src.holder || !holder.CheckAdminHref(href, href_list))
+		return
+	var/target = GET_VV_TARGET
+	vv_do_basic(target, href_list, href)
+	if(isdatum(target))
+		var/datum/D = target
+		D.vv_do_topic(href_list)
+	else if(islist(target))
+		vv_do_list(target, href_list)
+	if(href_list["Vars"])
```
Comment (vvvv-vvvv) [inline code/modules/admin/view_variables/topic.dm:13]: you need to change hrefs where it was `vars=` to `Vars=`

**Review by vvvv-vvvv — CHANGES_REQUESTED** [review]

File: code/__DEFINES/vv.dm [hunk 2]
```diff
@@ -16,26 +18,146 @@
...
+// /atom
+#define VV_HK_ATOM_JUMP_TO "atom_jump_to"
+#define VV_HK_MODIFY_TRANSFORM "atom_transform"
+#define VV_HK_ADD_REAGENT "addreagent"
+#define VV_HK_MODIFY_FILTERS "modify_filters"
+#define VV_HK_MODIFY_GREYSCALE_COLORS "modify_greyscale_colors"
+#define VV_HK_MODIFY_GREYSCALE "modify_greyscale"
+#define VV_HK_AUTO_RENAME "auto_rename"
+
+#define VV_HK_SPIN_ANIMATION "atom_spin" //Unused, from TG
```
[the remark this answers is not in the document]
Comment (Xander3359) [reply code/__DEFINES/vv.dm:97]: Bit out of scope for now.
There's a lot more TG stuff I want to work on in a followup

File: code/__HELPERS/_planes.dm [hunk 3]
```diff
@@ -0,0 +1,88 @@
+// This file contains helper macros for plane operations
```
[the remark this answers is not in the document]
Comment (Xander3359) [reply code/__HELPERS/_planes.dm:1]: I think this was multi-z, which again, I'll probably start working on in a followup

File: code/__DEFINES/_helpers.dm [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// Stuff that is relatively "core" and is used in other defines/helpers
+
+/*//Returns the hex value of a decimal number
+//len == length of returned string
+#define num2hex(X, len) num2text(X, len, 16)
+
+//Returns an integer given a hex input, supports negative values "-ff"
+//skips preceding invalid characters
+#define hex2num(X) text2num(X, 16)*/
```
[the remark this answers is not in the document]
Comment (Xander3359) [reply code/__DEFINES/_helpers.dm:9]: I did try, I simply don't understand how, and after having asked for help in multiple places, I never really got an answer. I don't know anything about hex so I left it here for someone who is more competent than I to implement it.

**Comment by TiviPlus** [comment]
fell free to ping me when you need reviews for this btw

File: code/__DEFINES/_helpers.dm [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+/// It is slightly less optimal then a []'d datum, but the cost is massively outweighed by the potential savings
+/// It will only work for datums mind, for datum reasons
+/// : because of the embedded typecheck
+#define text_ref(datum) (isdatum(datum) ? (datum:cached_ref ||= "\ref[datum]") : ("\ref[datum]"))
+#endif
+
+// Refs contain a type id within their string that can be used to identify byond types.
+// Custom types that we define don't get a unique id, but this is useful for identifying
+// types that don't normally have a way to run istype() on them.
+#define TYPEID(thing) copytext(REF(thing), 4, 6)
```
Comment (ivanmixo) [inline code/__DEFINES/_helpers.dm:43]: https://github.com/tgstation/TerraGov-Marine-Corps/blob/e3391cee330686cbacc80eeb32ebedc856dc8f5f/code/__DEFINES/typeids.dm#L5

File: code/__DEFINES/vv.dm [hunk 6]
```diff
@@ -16,26 +18,135 @@
...
+#define VV_MSG_MARKED "<br><font size='1' color='red'><b>Marked Object</b></font>"
+#define VV_MSG_TAGGED(num) "<br><font size='1' color='red'><b>Tagged Datum #[num]</b></font>"
+#define VV_MSG_EDITED "<br><font size='1' color='red'><b>Var Edited</b></font>"
+#define VV_MSG_DELETED "<br><font size='1' color='red'><b>Deleted</b></font>"
+
+#define VV_NORMAL_LIST_NO_EXPAND_THRESHOLD 50
+#define VV_SPECIAL_LIST_NO_EXPAND_THRESHOLD 150
+
+//#define IS_VALID_ASSOC_KEY(V) (istext(V) || ispath(V) || isdatum(V) || islist(V))
+#define IS_VALID_ASSOC_KEY(V) (!isnum(V)) //hhmmm..
```
Comment (ivanmixo) [inline code/__DEFINES/vv.dm:39]: Whar

File: code/__HELPERS/type2type.dm [hunk 7]
```diff
@@ -372,3 +295,24 @@
 /// for use inside of browse() calls to html assets that might be loaded on a cdn.
 /proc/url2htmlloader(url)
 	return {"<html><head><meta http-equiv="refresh" content="0;URL='[url]'"/></head><body onLoad="parent.location='[url]'"></body></html>"}
+
+//word of warning: using a matrix like this as a color value will simplify it back to a string after being set
```
Comment (ivanmixo) [inline code/__HELPERS/type2type.dm:299]: autodocing these would be cool

File: code/__HELPERS/type_processing.dm [hunk 8]
```diff
@@ -0,0 +1,66 @@
+/proc/make_types_fancy(list/types)
```
Comment (ivanmixo) [inline code/__HELPERS/type_processing.dm:1]: Autodoccing these would also be cool

File: code/modules/admin/view_variables/mark_datum.dm [hunk 9]
```diff
@@ -1,15 +1,19 @@
-/**
```
Comment (ivanmixo) [inline code/modules/admin/view_variables/mark_datum.dm:1]: Poor autodoc what'd he do to you :(

File: code/modules/admin/view_variables/mass_edit_variables.dm [hunk 10]
```diff
@@ -0,0 +1,274 @@
+/client/proc/cmd_mass_modify_object_variables(atom/A, var_name)
+	set category = "Debug"
+	set name = "Mass Edit Variables"
+	set desc="(target) Edit all instances of a target item's variables"
```
Comment (ivanmixo) [inline code/modules/admin/view_variables/mass_edit_variables.dm:4]:
```suggestion
	set desc = "(target) Edit all instances of a target item's variables"
```

File: code/__DEFINES/_helpers.dm [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+#ifdef EXPERIMENT_515_DONT_CACHE_REF
+/// Takes a datum as input, returns its ref string
+#define text_ref(datum) ref(datum)
+#else
+/// Takes a datum as input, returns its ref string, or a cached version of it
+/// This allows us to cache \ref creation, which ensures it'll only ever happen once per datum, saving string tree time
+/// It is slightly less optimal then a []'d datum, but the cost is massively outweighed by the potential savings
+/// It will only work for datums mind, for datum reasons
+/// : because of the embedded typecheck
+#define text_ref(datum) (isdatum(datum) ? (datum:cached_ref ||= "\ref[datum]") : ("\ref[datum]"))
```
Comment (ivanmixo) [inline code/__DEFINES/_helpers.dm:37]: Seems you have an older version of this, just make it equal to mainline tg again instead. 

https://github.com/tgstation/tgstation/blob/281e82d2237aba5beab54b5defc0bc0016018681/code/__DEFINES/_helpers.dm#L29

**Review by ivanmixo — CHANGES_REQUESTED** [review]
Good work besides these few things. Would really like it if stuff was autodocced even if you aren't the one who wrote the code, for consistency's sake. (and you're missing a lot of docs here)

**Review by TiviPlus — APPROVED** [review]

**Comment by MiniMeatwad** [comment]
"Set Species", "Jump To" dropdown menu options are broken.

**Comment by Xander3359** [comment]
> "Set Species", "Jump To" dropdown menu options are broken.

I'll look into it when I have some free time

**Pull request merged**